### DISCUSSION TITLE
L2ARC vdev sometimes perfectly hot before sleep (suspend) of FreeBSD, suddenly cold (after wake)

### DISCUSSION THREAD

**Discussion by grahamperrin** [openzfs/zfs#13478] [Q&A]
I have two L2ARC devices on USB, with a script to gracefully offline the two before sleep. 

Sometimes, after wake, with onlining automated by [zfsd(8)](https://www.freebsd.org/cgi/man.cgi?query=zfsd&sektion=8&manpath=FreeBSD), I find one of the two cold. 

Yesterday, for example, `/dev/gpt/cache-august` (GUID `0x6285db6c05d8b7f1`): 

* `1456` blocks restored at 10:14 <https://pastebin.com/raw/cqxqs2tD>
* `0012` blocks restored at 18:37 <https://pastebin.com/raw/3i061EAh>

```text
% grep "suspend at 20220518" /var/log/messages
May 18 09:37:48 mowa219-gjp4-8570p-freebsd acpi[94223]: suspend at 20220518 09:37:48
May 18 17:45:32 mowa219-gjp4-8570p-freebsd acpi[23894]: suspend at 20220518 17:45:32
% 
```

```text
% grep resume /var/log/messages | grep kernel | grep May\ 18
May 18 10:14:05 mowa219-gjp4-8570p-freebsd kernel: (ada0:ahcich0:0:0:0): resume
May 18 18:37:15 mowa219-gjp4-8570p-freebsd kernel: (ada0:ahcich0:0:0:0): resume
% 
```

Please: is it normal for each L2ARC vdev to occasionally go cold (when taken online)? Or should I investigate further, as a possible bug?

----

I see [a 2020 comment](https://github.com/openzfs/zfs/discussions/11209#discussioncomment-232924) from @behlendorf: 

> …  The L2ARC is implemented as an on-disk ring buffer which is why eventually all of the block were overwritten. …

– but (sorry) I'm not sure that I understand. For _ring buffer_, Wikipedia redirects to [circular buffer](https://en.wikipedia.org/wiki/Circular_buffer), with this sentence: 

> … A property of the circular buffer is that when it is full and a subsequent write is performed, then it starts overwriting the oldest data. …

So, I expect and understand overwrite (loss) of the _oldest_, but I'm surprised by the suddenness of hot to cold (like, losing _everything_ at once). 

----

Side note: in recent months I have `vfs.zfs.l2arc.noprefetch=0` in [sysctl.conf(5)](https://www.freebsd.org/cgi/man.cgi?query=sysctl.conf&sektion=5&manpath=FreeBSD).

**Comment by IvanVolosyuk** [comment]
It would be nice to specify kernel and zfs versions in the bug report. I believe there were some fixes https://github.com/openzfs/zfs/commit/ab8a8f07457831a21db888afcf48910ce3e9214a in persistent L2ARC, which fixed an issue where the l2arc can loose all the objects due to a race on write and restore during the import.

**Reply by grahamperrin** [reply]
Thanks. 

```text
% zfs --version
zfs-2.1.99-FreeBSD_gbc3f12bfa
zfs-kmod-2.1.99-FreeBSD_gbc3f12bfa
% grep -v \# /boot/loader.conf | grep zfs
zfs_load="YES"
openzfs_load="NO"
% uname -aKU
FreeBSD mowa219-gjp4-8570p-freebsd 14.0-CURRENT FreeBSD 14.0-CURRENT #11 main-n255588-01235012e5b-dirty: Thu May 12 21:45:43 BST 2022     root@mowa219-gjp4-8570p-freebsd:/usr/obj/usr/src/amd64.amd64/sys/GENERIC-NODEBUG amd64 1400058 1400058
% 
```

> [ab8a8f0](https://github.com/openzfs/zfs/commit/ab8a8f07457831a21db888afcf48910ce3e9214a) 

This is present in `main` in FreeBSD: 

<https://cgit.freebsd.org/src/commit/?id=ab8a8f07457831a21db888afcf48910ce3e9214a>

https://github.com/freebsd/freebsd-src/commit/ab8a8f07457831a21db888afcf48910ce3e9214a

**Reply by grahamperrin** [reply]
Defocusing from sleep/wake, note to self: 

[lines 8337–8343 of zfs/arc.c at master · openzfs/zfs · GitHub](https://github.com/openzfs/zfs/blob/master/module/zfs/arc.c#L8337-L8343)

```text
 * *) We reserve SPA_MINBLOCKSIZE of space at the start of each L2ARC device
 *    for our header bookkeeping purposes. This contains a device header,
 *    which contains our top-level reference structures. We update it each
 *    time we write a new log block, so that we're able to locate it in the
 *    L2ARC device. If this write results in an inconsistent device header
 *    (e.g. due to power failure), we detect this by verifying the header's
 *    checksum and simply fail to reconstruct the L2ARC after reboot.
```

**Comment by grahamperrin** [comment]
>  … surprised by the suddenness of hot to cold (like, losing _everything_ at once). …

Sometimes, I'd encounter the plummet affecting all L2ARC devices at the same time. In recent months, I use three devices (14 + 14 + 7.2 G).

I might broaden the subject of this discussion, because sometimes (often?) the plummets were also found after starting or restarting the OS. I typically restart only if I upgrade packages in a new ZFS boot environment, then activate the environment before restarting the OS. 

----

Until recently, my habit was physical reconnection of devices before waking the computer. Assuming that presence would reduce the likelihood of things plummeting. 

Maybe a false assumption; recently, I choose to wait until a few seconds after wake, then reconnect the devices, one by one. If I'm not mistaken: **fewer plummets** 👍

Whilst the sense of things improving is not yet shown in e.g. Netdata: 

![image](https://user-images.githubusercontent.com/192271/233778169-a4863dbe-ba00-4c37-a21b-041557d0c792.png)

– I hope for things to be more obviously improved by me also breaking the habit of having the devices connected at computer start (or restart) time. 

I'll now experiment, for a few days, with **not** connecting the devices until shortly after the deamon starts …

**Reply by grahamperrin** [reply]
> … – I hope for things to be more obviously improved by me also breaking the habit of having the devices connected at computer start (or restart) time.
> 
> I'll now experiment, for a few days, with **not** connecting the devices until shortly after the deamon starts …

First test result: failure. All cold. I'll **hide this comment as off-topic**, might add further test results without un-hiding the comment. 

----

```text
% zpool status -x
all pools are healthy
% zpool iostat -v
                       capacity     operations     bandwidth 
pool                 alloc   free   read  write   read  write
-------------------  -----  -----  -----  -----  -----  -----
august               711G   201G   85     32     1.84M  692K 
  ada0p3.eli         711G   201G   85     32     1.84M  692K 
cache                    -      -      -      -      -      -
  gpt/cache2-august  202M   14.2G  0      1      1.14K  505K 
  gpt/cache-august   337M   14.1G  0      1      1.14K  845K 
  gpt/cache3-august  206M   7.02G  0      1      1.14K  515K 
-------------------  -----  -----  -----  -----  -----  -----
% uptime
12:06p.m.  up 7 mins, 2 users, load averages: 0.12, 0.17, 0.08
% 
```

In this case, I connected the three devices shortly after ttyv0 showed the start of SDDM. 

From `/var/log/console.log`: 

`Apr 22 12:00:46 mowa219-gjp4-8570p-freebsd kernel: Starting sddm.`

From `/var/log/messages`: 

```text
Apr 22 12:00:31 mowa219-gjp4-8570p-freebsd kernel: GEOM_ELI: Device ada0p2.eli created.
Apr 22 12:00:31 mowa219-gjp4-8570p-freebsd kernel: GEOM_ELI: Encryption: AES-XTS 128
Apr 22 12:00:31 mowa219-gjp4-8570p-freebsd kernel: GEOM_ELI:     Crypto: accelerated software
Apr 22 12:00:55 mowa219-gjp4-8570p-freebsd apps.plugin[2864]: PROCFILE: Cannot open file '/usr/local/etc/netdata/apps_groups.conf'
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd kernel: ugen0.4: <Kingston DataTraveler 3.0> at usbus0
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd kernel: umass0 on uhub1
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd kernel: umass0: <Kingston DataTraveler 3.0, class 0/0, rev 2.10/1.00, addr 3> on usbus0
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd kernel: da0 at umass-sim0 bus 0 scbus3 target 0 lun 0
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd kernel: da0: <Kingston DataTraveler 3.0 1.01> Removable Direct Access SPC-4 SCSI device
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd kernel: da0: Serial Number 60A44C426697AF20F000167F
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd kernel: da0: 40.000MB/s transfers
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd kernel: da0: 14832MB (30375936 512 byte sectors)
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd kernel: da0: quirks=0x2<NO_6_BYTE>
Apr 22 12:00:58 mowa219-gjp4-8570p-freebsd ZFS[3009]: vdev state changed, pool_guid=1913339710710793892 vdev_guid=7383378062915565797
Apr 22 12:01:00 mowa219-gjp4-8570p-freebsd apps.plugin[2864]: Received error on stdin.
Apr 22 12:01:00 mowa219-gjp4-8570p-freebsd kernel: pid 2580 (netdata), jid 0, uid 302: exited on signal 11 (core dumped)
Apr 22 12:01:05 mowa219-gjp4-8570p-freebsd root[3027]: /etc/rc: WARNING: failed to start sendmail_submit
Apr 22 12:01:07 mowa219-gjp4-8570p-freebsd kernel: ugen0.5: <Kingston DataTraveler 3.0> at usbus0
Apr 22 12:01:09 mowa219-gjp4-8570p-freebsd kernel: umass1 on uhub6
Apr 22 12:01:09 mowa219-gjp4-8570p-freebsd kernel: umass1: <Kingston DataTraveler 3.0, class 0/0, rev 3.00/1.10, addr 4> on usbus0
Apr 22 12:01:09 mowa219-gjp4-8570p-freebsd kernel: da1 at umass-sim1 bus 1 scbus4 target 0 lun 0
Apr 22 12:01:09 mowa219-gjp4-8570p-freebsd kernel: da1: <Kingston DataTraveler 3.0 PMAP> Removable Direct Access SPC-4 SCSI device
Apr 22 12:01:09 mowa219-gjp4-8570p-freebsd kernel: da1: Serial Number 08606E6B6446BFB138159554
Apr 22 12:01:09 mowa219-gjp4-8570p-freebsd kernel: da1: 400.000MB/s transfers
Apr 22 12:01:09 mowa219-gjp4-8570p-freebsd kernel: da1: 14784MB (30277632 512 byte sectors)
Apr 22 12:01:09 mowa219-gjp4-8570p-freebsd kernel: da1: quirks=0x2<NO_6_BYTE>
Apr 22 12:01:09 mowa219-gjp4-8570p-freebsd ZFS[3199]: vdev state changed, pool_guid=1913339710710793892 vdev_guid=14605366699092709457
Apr 22 12:01:29 mowa219-gjp4-8570p-freebsd kernel: ugen0.6: <Kingston DataTraveler 2.0> at usbus0
Apr 22 12:01:29 mowa219-gjp4-8570p-freebsd kernel: umass2 on uhub3
Apr 22 12:01:29 mowa219-gjp4-8570p-freebsd kernel: umass2: <Kingston DataTraveler 2.0, class 0/0, rev 2.00/1.00, addr 5> on usbus0
Apr 22 12:01:29 mowa219-gjp4-8570p-freebsd kernel: da2 at umass-sim2 bus 2 scbus5 target 0 lun 0
Apr 22 12:01:29 mowa219-gjp4-8570p-freebsd kernel: da2: <Kingston DataTraveler 2.0 PMAP> Removable Direct Access SPC-2 SCSI device
Apr 22 12:01:29 mowa219-gjp4-8570p-freebsd kernel: da2: Serial Number 60A44C4253CABDA03B46026A
Apr 22 12:01:29 mowa219-gjp4-8570p-freebsd kernel: da2: 40.000MB/s transfers
Apr 22 12:01:29 mowa219-gjp4-8570p-freebsd kernel: da2: 7396MB (15147008 512 byte sectors)
Apr 22 12:01:29 mowa219-gjp4-8570p-freebsd kernel: da2: quirks=0x3<NO_SYNC_CACHE,NO_6_BYTE>
Apr 22 12:01:30 mowa219-gjp4-8570p-freebsd ZFS[3209]: vdev state changed, pool_guid=1913339710710793892 vdev_guid=10071853212821788401
```

### Postscript

This is odd: 

![image](https://user-images.githubusercontent.com/192271/233783490-03b7603d-de8d-4e93-b8c8-da49430fbfae.png)

Compare that with the screenshot in the previous comment, taken around ninety minutes earlier, when the Netdata view of things had no gap at the far right of the timeline. 

Shutdown was around 11:58 (BOOT was at 12:00:28), now there appears to be a gap of around three hours. 

Tentative cross-reference: FreeBSD bug [270997 – net-mgmt/netdata: sometimes not running when expected, sometimes running but http://localhost:19999 not responding](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=270997)

**Reply by grahamperrin** [reply]
> … I'll **hide this comment as off-topic**, …

Sorry, I discovered that hiding one's own commentary is a feature of issues but not discussions.
